pyvrp/educate/tests/test_Exchange.py

```diff
@@ -0,0 +1,190 @@
+    pm = PenaltyManager(data.vehicle_capacity)
+    rng = XorShift128(seed=42)
+
+    params = LocalSearchParams(nb_granular=data.num_clients)
+    ls = LocalSearch(data, pm, rng, params)
+    op = operator(data, pm)
+    ls.add_node_operator(op)
+
+    single_route = list(range(1, data.num_clients + 1))
+    individual = Individual(data, pm, [single_route])
+    copy = Individual(individual)
+
+    ls.search(individual)
+
+    # The new solution should strictly improve on our original solution, and
+    # should use more routes.
+    assert_(individual.num_routes() > 1)
+    assert_(individual.cost() < copy.cost())
+
+
+@mark.parametrize(
+    "operator",
+    [
+        Exchange22,
+        Exchange30,
+        Exchange31,
+        Exchange32,
+        Exchange33,
+    ],
+)
+def test_cannot_exchange_when_parts_overlap_with_depot(operator):
+    """
+    (N, M)-exchange works by exchanging N nodes starting at some node U with
+    M nodes starting at some node V. But when there is no sequence of N or M
+    nodes that does not contain the depot (because the routes are very short),
+    then no exchange is possible.
+    """
+    data = read("data/OkSmall.txt")
+    pm = PenaltyManager(data.vehicle_capacity)
+    rng = XorShift128(seed=42)
+
+    params = LocalSearchParams(nb_granular=data.num_clients)
+    ls = LocalSearch(data, pm, rng, params)
+    op = operator(data, pm)
+    ls.add_node_operator(op)
+
+    individual = Individual(data, pm, [[1, 2], [3], [4]])
+    copy = Individual(individual)
+
+    ls.search(individual)
+
+    assert_equal(individual, copy)
+
+
+@mark.parametrize("operator", [Exchange32, Exchange33])
+def test_cannot_exchange_when_segments_overlap(operator):
+    """
+    (3, 2)- and (3, 3)-exchange cannot exchange anything on a length-four
+    single route solution: there's always overlap between the segments.
+    """
+    data = read("data/OkSmall.txt")
+    pm = PenaltyManager(data.vehicle_capacity)
+    rng = XorShift128(seed=42)
+
+    params = LocalSearchParams(nb_granular=data.num_clients)
+    ls = LocalSearch(data, pm, rng, params)
+    op = operator(data, pm)
+    ls.add_node_operator(op)
+
+    individual = Individual(data, pm, [[1, 2, 3, 4]])
+    copy = Individual(individual)
+
+    ls.search(individual)
+
+    assert_equal(individual, copy)
+
+
+def test_cannot_swap_adjacent_segments():
+    """
+    (2, 2)-exchange on a single route cannot swap adjacent segments, since
+    that's already covered by (2, 0)-exchange.
+    """
+    data = read("data/OkSmall.txt")
+    pm = PenaltyManager(data.vehicle_capacity)
+    rng = XorShift128(seed=42)
+
+    params = LocalSearchParams(nb_granular=data.num_clients)
+    ls = LocalSearch(data, pm, rng, params)
+    op = Exchange22(data, pm)
+    ls.add_node_operator(op)
+
+    individual = Individual(data, pm, [[1, 2, 3, 4]])
+    copy = Individual(individual)
+
+    ls.search(individual)
+
+    assert_equal(individual, copy)
```

Comment:
Are you sure [3, 4, 1, 2] is an improvement over [1, 2, 3, 4], otherwise that may be the reason the solution is unchanged?

Reply:
I thought so, but I'll check to be sure and make this a bit more explicit.